File: webui/admin/itemList/itemListProps/userRole/listPrivs.js
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
function listRolePrivsRoleNameHTML(roleName) {
	return '' + 
		'<div class="row">' +
			'<label>' + roleName + '</label>' +
		'</div>';
}

function listPropsPrivsButtonRowHTML(roleID) {
	return '' + 
		'<div class="row formRolePrivsPrivRadioRow">' +
			formRolePrivsButtonsHTML(roleID) + 
		'</div>';
}


function listRolePrivilegeListItemHTML(listPriv) {
		
	return '' +
		'<div class="list-group-item formRolePrivListItem" id="'+listPriv.roleID+'">' +
			'<div class="container-fluid">' +
				listRolePrivsRoleNameHTML(listPriv.roleName) +
				formRolePrivsButtonsHTML(listPriv.roleID)
			'</div>' +
		'</div>';
}

function initListRolePrivProperties(listID) {
	
	jsonAPIRequest("userRole/getListRolePrivs", { listID: listID }, function(listPrivs) {
		
		var $privList = $('#listRolesPrivilegesList')
		
		$privList.empty()
		
		for(var listPrivIndex=0; listPrivIndex < listPrivs.length; listPrivIndex++) {
			var listPriv = listPrivs[listPrivIndex]
			$privList.append(listRolePrivilegeListItemHTML(listPriv))
			
			initFormRolePrivsButtons(listPriv.roleID,listPriv.privs, function(roleID,privs) {
				var setListRolePrivParams = {
					listID: listID,
					roleID: roleID,
					privs: privs
				}
				jsonAPIRequest("userRole/setListRolePrivs", setListRolePrivParams, function(listPrivs) {
					console.log("Updating list privileges: done")			
				})
			})	
		}
	})
}

File: server/userRoleController/listPrivs.go
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
package userRoleController

import (
	"database/sql"
	"fmt"
	"github.com/resultra/resultra/server/itemList"
	"github.com/resultra/resultra/server/userRole"
)

func getRoleListPrivsWithDefaults(trackerDBHandle *sql.DB, roleID string) ([]userRole.RoleListPriv, error) {

	roleDB, err := userRole.GetUserRoleDatabaseID(trackerDBHandle, roleID)
	if err != nil {
		return nil, fmt.Errorf("getRoleListPrivsWithDefaults: %v", err)
	}

	privsByListID := map[string]userRole.RoleListPriv{}

	// Start off with no privileges as the default for all lists

	allItemLists, err := itemList.GetAllItemLists(trackerDBHandle, roleDB)
	if err != nil {
		return nil, fmt.Errorf("getRoleListPrivsWithDefaults: %v", err)
	}
	for _, currList := range allItemLists {
		privsByListID[currList.ListID] = userRole.RoleListPriv{
			ListID:   currList.ListID,
			ListName: currList.Name,
			Privs:    userRole.ListRolePrivsNone}
	}

	// Update the privileges for those lists with an explicit set of privileges set.
	explicitListPrivs, err := userRole.GetRoleListPrivs(trackerDBHandle, roleID)
	if err != nil {
		return nil, fmt.Errorf("getRoleListPrivsWithDefaults: %v", err)
	}
	for _, currPriv := range explicitListPrivs {
		privsByListID[currPriv.ListID] = currPriv
	}

	// Flatten the list
	roleListPrivs := []userRole.RoleListPriv{}
	for _, currPriv := range privsByListID {
		roleListPrivs = append(roleListPrivs, currPriv)
	}

	return roleListPrivs, nil

}
